### PR TITLE
[commhistory] Add a filter property to events.

### PR DESCRIPTION
**Pull request by dcaliste**
Since oFono has filtering capabilities,
expose an enum for events that can
represent these capabilities.

Adding such property allow oFono filter
plugins to expose to the user their
activities.

It is done by storing an additional
extra property for events and it provides
getters and setters for it.

@pvuorela , when you will have time, may we discuss of the possibility to add this new API to the comm history ? If you want to see it in action, you can give a look to https://github.com/dcaliste/sailfish-voicecall-filter (which provides an oFono plugin to filter incoming calls and a user daemon that dialogs with the oFono plugin via D-Bus and applies a simple filtering rule from DConf lists of ignored or blocked numbers). To get a feedback in the UI, you can add the following lines in `/usr/share/voicecall-ui-jolla/common/CallDirectionIcon.qml` (in the icon entry):
```
            } else if (call.filterType !== CommCallModel.NoFilter) {
                return "image://theme/icon-s-blocked"
```

The rational to add such enum in the comm history, is that contrary to an MDM filter plugin (which should stay silent), the user would like to monitor when calls were blocked or ignored (at least I do). And the logical place to look at is the call history. Doing it like in this PR via an extra property of Event won't affect existing API of course and don't require any change on the database layout. So it's quite light weight. As for the changes in the UI (the blocked icon even already exists !).

### CODE REVIEW

**Comment by pvuorela** [comment]
Don't have much time left today, but as quick comment for this side I'm pondering whether the enums are a bit mixing up filter as in 'an incoming call was caught by a filter' and a filter as in 'please list me events filtered by this rule'.

In general we should look into implementing the black list feature in sailfish os. For that I wouldn't like to have extra daemons or ofono plugins.

But need to get goign now...

**Comment by dcaliste** [comment]
> I'm pondering whether the enums are a bit mixing up filter as in 'an incoming call was caught by a filter' and a filter as in 'please list me events filtered by this rule'.

Yes, good remark. We may keep "filter" verb for model filtering. The enum may be called `EventReceptionStatus` or something in that direction. Which brings me to another possibility I thought about : why not add two new status in the already existing `EventStatus` enum ? It could make sense also. What do you think ?

> In general we should look into implementing the black list feature in sailfish os.

I agree (for whatever it's worth). That's why I'm proposing here some (minor) additions to the core elements like the comm history. I've some modifications to propose to the UI parts also (sailfish-components-contacts for instance). I was just testing the "water temperature" about OS inclusion with this PR here…

> For that I wouldn't like to have extra daemons or ofono plugins.

May I ask why ? The architecture in oFono for call blocking is quite nice. Every previous attempt to react on D-Bus incoming call  in user session and reply with an ignore or hangup D-Bus request, ended up with sometime the phone ringing once, as far as I know. While, with an oFono plugin, the broadcast of the icoming call is delayed up to the filter plugin answer. The drawback for the oFono plugin is that it lives in root session and thus requires to speak to user via a daemon there, as @monich mentioned. As you can see in my attempt to write such a daemon, it can be very light weight (a D-Bus proxy and a filtering class).

If you agree, nothing is in a hurry and we can take time to discuss the best ways to get blacklist (and whitelist) for incoming blocking and user feedback and interface. Anyway, thanks for considering this.

**Comment by pvuorela** [comment]
>  Which brings me to another possibility I thought about : why not add two new status in the already existing EventStatus enum ? It could make sense also. What do you think ?

At the moment Event::EventStatus is quite much about message delivery status. Adding blocking there wouldn't immediately feel like a good idea.

> I agree (for whatever it's worth). That's why I'm proposing here some (minor) additions to the core elements like the comm history. I've some modifications to propose to the UI parts also (sailfish-components-contacts for instance). I was just testing the "water temperature" about OS inclusion with this PR here…

I think the minimal support would be adding blocked call listing to Phone app settings which includes already something on that direction. Slightly better having that available also on contact page. Whether the blocked calls are listed and how feels like second phase thing. But per se it would be nice having support for blocking.

Btw, I didn't quite catch what's the difference between blocked and ignored.

> > For that I wouldn't like to have extra daemons or ofono plugins.

> May I ask why ? 

On extra daemons part just to avoid memory usage and there existing already various daemons for call handling. Though I suppose if there was an ofono plugin, the user specific code in your daemon could be merged into one of those existing daemons. On ofono side below.

> The architecture in oFono for call blocking is quite nice. Every previous attempt to react on D-Bus incoming call in user session and reply with an ignore or hangup D-Bus request, ended up with sometime the phone ringing once, as far as I know. While, with an oFono plugin, the broadcast of the icoming call is delayed up to the filter plugin answer. The drawback for the oFono plugin is that it lives in root session and thus requires to speak to user via a daemon there, as @monich mentioned. As you can see in my attempt to write such a daemon, it can be very light weight (a D-Bus proxy and a filtering class).

I was thinking it would be nice having the call specific logic more towards voicecall/voicecall-jolla to limit how widely it spreads. Also the UI for the blocking goes towards this direction. For one it would be nice if it didn't need to expose the configuration too much externally. 

Also, as you mention about root session and user specific parts, connecting from former to latter can be problematic. Looks like you're now doing some signal and expecting a method call to be done soon enough(?).

Maybe related to the "ringing once", earlier the voicecall-qt5 started playing the ringtone but with do-not-disturb and sim2 ringtone setting I made the voicecall-jolla more in control on when the ringtone starts playing. I think it should be possible to do that only after it's known whether the number is blocked.

Then Ofono upstream has been moving towards removing plugin support. We are not likely able to have such a version, but still good to consider if we want to introduce any further plugins.

**Comment by dcaliste** [comment]
> Slightly better having that available also on contact page. Whether the blocked calls are listed and how feels like second phase thing.

In my opinion, these are two unrelated things and can be worked in parallel:
* you need a place to handle how to decide which number(s) should be blocked,
* and you need a feedback to be able to supervise what the device is doing.

My initial thoughts on the first point was to add a pulley menu action for unsaved contacts, that adds this number to the blocked list (which switch to unblock when this is a blocked number). I guess no one wants to block a known contact… A secondary text in the unsaved contact card can mentioned that this number is blocked. It's simple to access this from the call list : you receive a phone call. You understand it's spammers, and then you open the contact card and block it for next time. This UI design sadly is not working well with spammers that changes slightly their number each time. You need a wild card for this… But I've currently no idea how to handle wild cards nicely in the UI…

The second point is to be able to monitor what is happening and that's where the call history enters into play. It's a parallel addition. I agree that you first need to be able to add new numbers to the block list to be able to see it in the call history ;-)

The developments I've started are turning around a library that handle number matching… The idea is also to use this library to expose QML object that can be used in the UI. Since, they are more invasive changes, I started with the call history which didn't require to have much changes. But if you think it's worth, I can continue the QML development in the library and propose some PR into sailfish-component-contacts.

> Btw, I didn't quite catch what's the difference between blocked and ignored.

That's a difference on how you handle the incoming call:
* when you ignore, you just let the incoming call continue, but don't ring and at the end for the caller, it's as if you didn't reply,
* or when you block, the incoming call is actively hang up and the caller is sent to the voicemail or whatever is configured with your provider.

As a first implementation, I don't think it's a difference that should be exposed in the UI. Ignore strategy is plainly enough for spammers, I guess.

> Looks like you're now doing some signal and expecting a method call to be done soon enough(?).

Exactly. That's how it is working in oFono as far as I undertand it. The voicecallfilter.c, if any filtering plugin exists, is waiting for any plugin answer to continue propagating the call to D-Bus and the rest of the system, or to actively block or ignore it. The system - user dialog is done in my plugin as a service on oFono side, emiting a signal, asking for filtering. Then, the user side is replying, if not in a given amount of time, the plugin let the call continues normally.

> I was thinking it would be nice having the call specific logic more towards voicecall/voicecall-jolla to limit how widely it spreads. 

So you suggest for me to have a look into `voicecall-qt5` instead of the (Sailifsh-) oFono plugin route ? I can add the filtering logic there, I guess, based on that library I was talking about before.

> For one it would be nice if it didn't need to expose the configuration too much externally.

That's the idea of the library here. It provides an API (C++ and QML objects) to know if a number is blocked, adding or removing a number from the block list… But internally, have full control without exposing it on how we store / implement the block numbers.

**Comment by pvuorela** [comment]
> I guess no one wants to block a known contact…

I haven't gotten a need to block anyone, but if I did, I could imagine doing it on a saved contact. That way I'd have some reference on what the number is about instead of later looking into list of blocked numbers and wondering what they are about.

>This UI design sadly is not working well with spammers that changes slightly their number each time. You need a wild card for this… But I've currently no idea how to handle wild cards nicely in the UI…

Wild card sounds relatively advanced and would require the user figuring out the pattern that's used. If there even is one.

I don't think this is phase 1 feature.

What I'd maybe start with is just a simple UI on Phone app settings page under call barring. Ability there to add and remove phone numbers that get blocked.

> So you suggest for me to have a look into voicecall-qt5 instead of the (Sailifsh-) oFono plugin route ? I can add the filtering logic there, I guess, based on that library I was talking about before.

I'd expect voicecall-jolla being even better place. And if it has the settings page + block evaluation, the configuration does not yet need to be exposed elsewhere.

**Comment by pvuorela** [comment]
For naming, how about something like renaming EventFilterType -> EventBlock(ed?)Status, NoFilter -> NotBlocked, and so on. Avoiding possibility to mix up list filtering now  and ignoring call that happened somewhere in the past.

Needs a rebase now.

**Comment by dcaliste** [comment]
Thanks for the naming proposition. It brings me a suggestion : why not use this new enum to give details on incoming call status ? Like there is a `isRead` flag complemented by a `eventReadStatus`, I'm proposing to expand the `isMissedCall` flag. Actually, incoming calls have now two status : received and missed. Adding call filtering will add two new status : blocked and ignored. So I'm proposing to cretae a new enum for incoming status. Since it is partially redundant with isMissedCall flag, it requires to keep both in sync, thus some additional lines in the setters. But doing so would allow easy labelling of incoming calls in UI and avoid constructions like `if !isMissed && filterType == Blocked then ...`.

I'm not completely sure, so it's just an additionnal commit on top. I can either squash it or drop it for what you were proposing (EventBlockedStatus).

File: src/event.h
Comment (pvuorela) [inline src/event.h]: Leftover?

Comment (dcaliste) [reply src/event.h]: Good catch. It's actually unused. I've removed it.

File: src/event.cpp
Comment (pvuorela) [inline src/event.cpp]: The existing missed calls won't be having the extra property so could use the flag for marking the event missed? It could use both extra property or the flag, but maybe could just keep the missed call state storage as it was to avoid redundant and potentially conflicting data?

**Comment by pvuorela** [comment]
I like this naming a lot more than the filter. Thanks for change. The commits could be squashed, though, to avoid going back and forth.

File: src/event.cpp
Comment (dcaliste) [reply src/event.cpp]: I'm doing the synchronisation in the setters:
- `setIsMissed()` is changing the status to Missed when true (left it unchanged otherwise),
- `setIncomingStatus()` is calling setIsMissed(status == Missed).

Like that, if I'm not mistaken, they should always be consistent. For older database entries, `setIsMissed()` is called from `DatabaseIOPrivate::readEventResult()`, which should set the proper incoming status.

Comment (dcaliste) [reply src/event.cpp]: Ah, but I see your point of not saving "missed" as an extra property, not to be redundant and potentially faulty.

I've amended the commit accordingly, to save only the two new states.

**Review by pvuorela — APPROVED** [review]
Looks good to me now. 

Not sure whether to merge already. So far haven't really tested the new statuses as there's no code actually using it. Then again shouldn't break anything else, and if there would be problems, they should be surfacing if something using these was attempted.

File: src/event.cpp
Comment (pvuorela) [inline src/event.cpp]: Infinite loop here. setIncomingStatus() calls setIsMissedCall()

Comment (dcaliste) [reply src/event.cpp]: Thank you, it will save me some debugging time.

**Comment by dcaliste** [comment]
I've uniformised the reject action (when an incoming call is hang up automatically on arriving), sometimes I call it "blocked", sometimes "rejected". I've changed everywhere here for "rejected".

**Comment by pvuorela** [comment]
Btw one thing I've kept pondering is whether a blocked incoming call is "missed" or not. One could argue both ways, it's missed in the sense that it wasn't answered (should isMissedCall=false be considered answered?), but then again not missed in the sense that it was intentionally ignored.

On a practical level it is now used in voicecall-jolla for call duration UI as something like "visible: !model.isMissedCall" which would go wrong on blocking calls. Throughout the system it's not that many lines and need to check all of them anyway to handle the new states, but blocked being also missed might be slightly more backwards compatible. Not sure how much it matters, though.

**Comment by dcaliste** [comment]
> Btw one thing I've kept pondering is whether a blocked incoming call is "missed" or not.

You're right. For instance, the current proposition is breaking the filtering of event models. If you choose ReceivedCall as a filter, you'll also get the ignored and blocked ones because they don't have the isMissedCall flag. And it's quite cumbersome to set a proper query specifying `and if there is no incomingStatus in the extraProperties, or if the value of this incomingStatus property is "ignored" or "rejected"`. The best solution would be to set the isMissedCall flag for non answered calls but also for ignored or rejected ones. Like that, the filtering options still wrks, assuming that ignored or rejected calls are missed ones. The grouping capability in callmodel.cpp can then be updated to work on incoming status instead of solely isMissedCall flag. I'm going to try such an approach and see what it breaks.

**Review by pvuorela — APPROVED** [review]
Works nice now with the voicecall PR. Thanks!

I'll start by having this in already now.

**Comment by dcaliste** [comment]
Thank you, I forgot to mention that I changed the IncomingStatus::Missed for IncomingStatus::NotAnswered. Like that, the isMissedCall flag can be used for all non received calls, and then the IncomingStatus can give details about the reason for not receiving a call. `IncomingStatus::Missed` was a nice wording, but it clashed with the isMissedCall flag which also encompass the ignored and rejected calls. I hope, you noticed the change and agreed with it !

**Pull request merged**
